### PR TITLE
Cast to fp16 before moving to device with deepspeed

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Blue: master
Orange: PR
Screenshot 2022-08-03 at 6 12 32 PM

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @tchaton @rohitgr7 @awaelchli @akihironitta @carmocca @justusschock

### CODE REVIEW

**Review by carmocca — APPROVED** [review]

**Comment by carmocca** [comment]
IMO this should be part of a bug-fix release

**Comment by rohitgr7** [comment]
> IMO this should be part of a bug-fix release

Then, should I move out the deprecated module from here in a sep PR or just say.. deprecated in 1.7.1 and will be removed in 1.9?

**Comment by carmocca** [comment]
I would choose the "deprecated in 1.7.1" option, but the other would also be fine.

File: src/pytorch_lightning/strategies/ipu.py [hunk 1]
```diff
@@ -269,7 +269,7 @@ def to_tensor(x):
         args = apply_to_collection(args, dtype=(int, float), function=to_tensor)
         return args
 
-    def batch_to_device(self, batch: Any, device: Optional[torch.device] = None, dataloader_idx: int = 0) -> Any:
+    def batch_to_device(self, batch: Any, *_: Any, **__: Any) -> Any:
```
Comment (justusschock) [inline src/pytorch_lightning/strategies/ipu.py:272]: can we still keep the signature as is? Even though these args may not be used, they should be part of the signature IMO as this really improves readability

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 2]
```diff
@@ -943,6 +944,19 @@ def register_strategies(cls, strategy_registry: Dict) -> None:
...
+        def fp_to_half(tensor: Tensor) -> Tensor:
+            if torch.is_floating_point(tensor):
+                if self.precision_plugin.precision == PrecisionType.HALF:
+                    return tensor.half()
+                elif self.precision_plugin.precision == PrecisionType.BFLOAT:
+                    return tensor.bfloat16()
+
+            return tensor
+
+        batch = apply_to_collection(batch, Tensor, function=fp_to_half)
```
Comment (justusschock) [inline src/pytorch_lightning/strategies/deepspeed.py:957]: should we move this to common utils? This is almost the same as with IPU and we should be able to exactly reuse that function as for IPU the precisiontype should always be HALF

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/strategies/deepspeed.py (hunk 2 above)
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/deepspeed.py]: done!

File: src/pytorch_lightning/strategies/ipu.py (hunk 1 above)
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/ipu.py]: reverted!

File: src/pytorch_lightning/strategies/utils.py [hunk 3]
```diff
@@ -13,6 +13,20 @@
...
+import torch
+
+from pytorch_lightning.utilities.enums import PrecisionType
+
 
 def on_colab_kaggle() -> bool:
     return bool(os.getenv("COLAB_GPU") or os.getenv("KAGGLE_URL_BASE"))
+
+
+def fp_to_half(tensor: torch.Tensor, precision: PrecisionType) -> torch.Tensor:
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/utils.py:25]: Make this protected

**Review by kaushikb11 — APPROVED** [review]

File: tests/tests_pytorch/strategies/test_deepspeed_strategy.py [hunk 4]
```diff
@@ -84,43 +84,6 @@ def automatic_optimization(self) -> bool:
         return False
 
 
-def test_deepspeed_lightning_module(tmpdir):
```
Comment (awaelchli) [inline tests/tests_pytorch/strategies/test_deepspeed_strategy.py:87]: we normally keep the tests and wrap them with a `deprecated_call()`, or move them to the test_remove_xxx

File: src/pytorch_lightning/strategies/utils.py [hunk 5]
```diff
@@ -13,6 +13,20 @@
 # limitations under the License.
 import os
 
+import torch
+
+from pytorch_lightning.utilities.enums import PrecisionType
+
 
 def on_colab_kaggle() -> bool:
```
Comment (awaelchli) [inline src/pytorch_lightning/strategies/utils.py:21]: Oh dang, we wanted to remove this and the file with it. 
Guess we forgot to deprecate it :( Will do it.

**Review by awaelchli — APPROVED** [review]

File: tests/tests_pytorch/strategies/test_deepspeed_strategy.py (hunk 4 above)
Comment (rohitgr7) [reply tests/tests_pytorch/strategies/test_deepspeed_strategy.py:87]: done!
